(3DS) Add bottom screen menu

bulzipke opened the pull request:
## Description
I have long hoped that experience similar to Virtual Console in 3DS with Retroarch.
Rather, the console log is being printed currently, so I thought it would be incomplete for gamers.

So, I PR but even a Retroarch Newbie.
I don't know if it fits the Retroarch style coding form, but please review the code.

Read the bottom menu png and draw it appear on 3DS bottom screen.
(assets/ctr/bottom_menu.png)

The following functions are currently available on bottom menu screen
Resume Game
 -> Toggle Retroarch menu
Create Restore Point
 -> If you touch this, save it to sd card.
 -> If the Save State Thumbnails option is enabled, you can see Screenshots at the Load Restore Point
Load Restore Point
 -> Shows the date you saved it. (And show Screenshot if exists)
 -> If you change the State Slot, bottom menu will refresh Screenshot / date / state.

Also I made sure to re-draw only when the bottom menu needs to be updated.
Battery time is precious when doing 3DS outdoors.

I completed the test on CTRULIB_2 and current toolchain.

And I added CONSOLE_LOG define to the Makefile.
If enable this, console log will print at bottom screen as before.
I wanted to allow this to be turned on and off as an option, but it appears to be an unsupported feature.
https://github.com/devkitPro/libctru/issues/476

(06/07) Save state to SRAM only, not disk. Write to disk when game stops or exits.
(06/26) Bottom texts use msg hash.

todo:
Improve of bottom_menu.png
 -> Unfortunately, I'm hopelessly not talented in this..
 -> The image I uploaded is from @hunterk by discord. Thank you for taking this opportunity.
 -> I want someone interested in this to help.
 -> and I've put PR on https://github.com/libretro/retroarch-assets/pull/408. Is this right process?
Improve the usage font of bottom screen.
-> I tried to change the font of bottom screen, but it was difficult to me.
-> Using System Font with CITRA2D. Failed.
-> Enable HAVE_STB_FONT. This does a great draw on the bottom screen but breaks the xmb theme.

## Related Pull Requests
https://github.com/libretro/retroarch-assets/pull/408



## Working video
https://user-images.githubusercontent.com/15259720/120219293-f73f1200-c275-11eb-9021-05bd1f061419.mp4

MrHuu commented:
That looks pretty nice! Exelent job. I've been working on something simmilar, adding a animated software keyboard and mouse, save states, rewind and fast forward to the bottom screen. It has the option to be disabled by holding, fading out when not in use and finally disabling the bottom screen backlight until a touch gesture is registered. i would love to help you out where i can. It will take a few days before i'm able to take a look though. Keep it up!

bulzipke commented:
> That looks pretty nice! Exelent job. I've been working on something simmilar, adding a animated software keyboard and mouse, save states, rewind and fast forward to the bottom screen. It has the option to be disabled by holding, fading out when not in use and finally disabling the bottom screen backlight until a touch gesture is registered. i would love to help you out where i can. It will take a few days before i'm able to take a look though. Keep it up!

Thank you for leaving comments!

To be honest, I don't know much about GPU / 3DS programming.
So if there's anything I need to fix about what I've done, please feel free to give me anything.

And I'm also interested in your animated bottom screen work!
Is it that can be merge with my work?

I think you know more about 3DS than I do.
If you need anything help, I'll :)

twinaphex commented:
Hi, when jdgleaver has some time, he will review this I think.

MrHuu commented:
A quick look through the PR reveals a lot of similarities, however the method of rendering differs.
I've been writing pre-converted (during compilation) assets written directly to one of the bottom screen framebuffers. If the screen needs updating, swap the framebuffers. When a key is pressed, overwrite the existing framebuffer with an other image at the right location.
I found this method to be pretty fast, requiring no loading times and not affecting performance to much.

I do like your method of using GPU textures a lot more to be honest. But unfortunately i'm not really familiar with manipulating GPU textures with lower level GSPGPU commands.

I'm still very much interested in having additional control options for the bottom screen. Merging the updated input_driver and adding the control functions would be easy, but the method of rendering would need to be adapted on my end.

If you wouldn't mind having a button added to pull up a keyboard or mouse to the existing layout, adding them should be pretty straightforward.


The implementation added with this PR is universal enough to be the default, but maybe an advanced ( or extra ) control option could be added as a retroarch toggle to also allow buttons for rewind / fast forward and bringing up the keyboard or mouse. This should probably be core dependent since not all cores will be able to use these functions.

This also brings up the question:
How does this PR behave when a port does not support serialization? dosbox-svn for example.

twinaphex commented:
Hi @justinweiss , since you were so instrumental in the 3DS port so far, would you like to take a look at this one too? If everything is fine with it and it's an overall improvement I'd love to be able to merge it.

bulzipke commented:
> A quick look through the PR reveals a lot of similarities, however the method of rendering differs.
> I've been writing pre-converted (during compilation) assets directly to one of the bottom screen framebuffers. If the screen needs updating, swap the framebuffers. When a key is pressed, overwrite the existing framebuffer with an other image at the right location.
> I found this method to be pretty fast, requiring no loading times and not affecting performance to much.
> 
> I do like your method of using GPU textures a lot more to be honest. But unfortunately i'm not really familiar with manipulating GPU textures with lower level GSPGPU commands.
> 
> I'm still very much interested in having additional control options for the bottom screen. Merging the updated input_driver and adding the control functions would be easy, but the method of rendering would need to be adapted on my end.
> 
> If you wouldn't mind having a button added to pull up a keyboard or mouse to the existing layout, adding them should be pretty straightforward.
> 
> The implementation added with this PR is universal enough to be the default, but maybe an advanced ( or extra ) control option could be added as a retroarch toggle to also allow buttons for rewind / fast forward and bringing up the keyboard or mouse. This should probably be core dependent since not all cores will be able to use these functions.
> 
> This also brings up the question:
> How does this PR behave when a core does not support serialization? dosbox-svn for example.

I couldn't think of a way to swap the framebuffer when need update time only.
Are pre-converted assets reading something like a bin file from romfs? I like XMB more than RGUI, and it would be great if it could be applied to that as well!
The bottom screen texture is being read on the sd card for easy theme customization now, but later changes to pre-converted assets will help disk io a lot. XD

I also referred to your overlay PR a lot in GPU commands. I pay tribute to your humility (_ _) 

As you said, there are many possibilities for the 3DS bottom screen.
I didn't even think of a keyboard and a mouse.
Also, I didn't think of a case that doesn't support serialization. It seems that it needs to be core-dependent deactivation or it looks good to see only your awesom keyboard and mouse that I'm looking forward to a lot!

As your advice, I should change to framebuffers swap only when there are need to refresh on the current work.


I have a different question from the subject. XD
Does the dosbox-svn run smoothly in 3DS?
I'm curious about compatibility now!

bulzipke commented:
I added the function SAVE/LODE STATETO RAM. And I changed this to the Bottom screen menu.
When closing the lid of a 3DS, pressing the HOME button, or exiting the Retroarch, the ram state is saved to disk.

I got a question while working on it.
Look at 00:27 in the video, thumbnail is broken. I'd like to ask for someone help on this.
I think it's wrong to get width/height from the code in the link below.
https://github.com/bulzipke/RetroArch/blob/3ds-bottom-screen/gfx/drivers/ctr_gfx.c#L522-L523

And I just found out, but I don't know why there's a build error :(

https://user-images.githubusercontent.com/15259720/120937354-f875b080-c747-11eb-9a51-0545f2fb8380.mp4

v1.9.4
00:04 ~ 00:15 : save state to disk
00:15 ~ 00:16 : load state from disk

This branch
00:22 ~ 00:22 : save state to ram
00:27 ~ 00:27 : load state from ram

bulzipke commented:
I did rebase my work. This is probably the last update.
Please call me need more update. Thank you! :)

twinaphex commented:
I believe @jdgleaver is still intending on checking this one out. Please bear with us for a while.

MarioKartFan commented:
Would absolutely love to see this make its way into the nightlies. Appears to be a huge improvement.

MrHuu commented:
While i see no direct issues, i do like to make a suggestion.

Most bottom screen related functions, although not directly related to graphics, are included in ```gfx/drivers/ctr_gfx.c```.

With the intend to add more functionality to the bottom screen, this goes out of the scope for the graphics driver itself.

As i would love to adapt the software keyboard / mouse i'm working on to this graphics implementation, little code is gfx related and is kept separate in the platform source folder ```ctr/ctr-bottom/ctr_bottom_kbd.c``` for example.

For this, the touch input polling will be done by the input driver calling the (selected mode) code.
Which in term signals the gfx driver to render on state change.

With this current pr, little room is left for additional features without separating the bottom screen functionality from the graphics driver itself first.


Please, do not take this as criticism as i can't wait to expand on the solid base provided here.
I just have to find the time to do it.

Also, this shouldn't be a reason to delay merging this functionality as is.

bulzipke commented:
@MrHuu Thank you for your detailed suggestion :)

You're right. My bottom screen code is concentrated in ctr_gfx.c.
As you suggested, the graphics and controller should be separated from the code.

I was interested in your work, so I went into your Retroarch github repository, but I couldn't find the branch.
(But there were interesting works such as ctr_romfs!)

I think your keyboard/mouse work is an excellent job.
I referred to the 3DS official Virtual Console interface on the bottom screen.
If you add a keyboard/mouse, there are will need a switch button. (I have no talent for UI..)
![ctr](https://user-images.githubusercontent.com/15259720/124749247-f2e5d300-df5e-11eb-8201-3b5f7c436040.png)

First of all, I will organize the code according to your suggestion when I have time.
Thanks once again!

MrHuu commented:
I haven't uploaded the code yet since it needs some refactoring and cleanup. Didn't expect anyone working on something similar to be honest.

This is what i came up with so far:
![20210709_171858](https://user-images.githubusercontent.com/561623/125103114-51848b80-e0dc-11eb-8e85-dfd813596537.jpg) ![20210709_171926](https://user-images.githubusercontent.com/561623/125103119-521d2200-e0dc-11eb-8060-19e76cd5c1fd.jpg)
![20210709_171907](https://user-images.githubusercontent.com/561623/125103117-521d2200-e0dc-11eb-8631-dadc7ca8ac84.jpg) ![20210709_171915](https://user-images.githubusercontent.com/561623/125103118-521d2200-e0dc-11eb-8880-a4c84f2141fc.jpg)
![20210709_172038](https://user-images.githubusercontent.com/561623/125103120-52b5b880-e0dc-11eb-9cd7-d52ce39e8f22.jpg)

I'll try and have it cleaned and upload it to git as reference. As most of the rendering code would be obsolete when merging with this pr, at least the input related code could still prove useful. But i think i'll work towards rendering using gpu textures right away, while at it.

I'm absolutely fine with the layout suggested above, the mouse could be accessible from the keyboard layout.
The assets could be adapted to be a bit more 'virtual console' like.

Any other features i had planned initially should be left alone for now.
The save sate functionality from this pr alongside keyboard / mouse support would suffice in my opinion.

The romfs branch saves you from having to copy thousands of files to your sd and improves loading times significantly. Especially when using the 'XMB' driver. But i'm not sure if it's suitable to use upstream, which is why i haven't submitted it.
The default asset directory has to be updated in both libretro-super and retroarch when using it merged upstream.

hizzlekizzle commented:
whew, that keyboard looks great :O

MrHuu commented:
![ctr_bottom_kbd_full](https://user-images.githubusercontent.com/561623/125163141-13df3b80-e18c-11eb-81d3-92e9f9418e44.png) ![ctr_bottom_kbd](https://user-images.githubusercontent.com/561623/125162997-50f6fe00-e18b-11eb-8afc-1864bd7c9fac.png)

Here's a quick mock-up on how it could look like. For accessibility (without using the stylus) i tried using the full screen, so maybe we should keep it that way instead?

MarioKartFan commented:
I love both of these but for ease of use, the layout that @bulzipke has shown above seems incredible. It would be nice to have a “best of both worlds” where the default was closer to @MrHuu and then when you tap on a “state button” you get the layout that @bulzipke presents.

Still, being eager to test this out would love to see this PR get merged. 

On an unrelated note if @MrHuu has solved the XMB load times on the 3DS that would be incredible.

bulzipke commented:
@MrHuu
Wow, your interface is very intuitive and Retroarch style :)
I think your work is great, too.

Maybe our code can be merge like this.
![bottom1](https://user-images.githubusercontent.com/15259720/125310964-02558980-e36e-11eb-991b-151684aaa848.png)
I think Plan A is expandable.
One remaining button might show a recently played game, or a thumbnail of a playlist, or console log, etc.
Plan B is just simple and easy to use.

We need a clear decision on which to choose.
(If you don't mind, I'd be happy to merge this PR to your branch)

MrHuu commented:
My initial plan was to have multiple features available: keyboard, mouse, save-states, and achievement information. Not necessarily limited to those.
Eventually selectable though a retroarch menu ( user interface -> bottom screen -> button1 = keyboard ).

If we're able to set the amount of buttons with their bound feature and save it as a core override, it's possible to only show kbd / mouse for dosbox and save-states for gpsp for example.

My personal preference goes to 'Plan A'.
With hardcoded features to start with.

I'm currently reworking loading / rendering textures in similar fashion as done in this pr.
With the keyboard nearly finished, i'll continue adding basic main functions and input driver.

I think it's best for me to create a pr to https://github.com/bulzipke/RetroArch/tree/3ds-bottom-screen and go from there, combining the two.
I should have a pr ready pretty soon.

twinaphex commented:
Hi there, I feel a bit sorry it's been two months since this PR was submitted.

How can we help you to fasttrack this and get it merged? What do you think is standing in the way from having this ready for release?

MrHuu commented:
Unfortunately, i won't be able to contribute in the near future. ( another 3 or 4 weeks )

While working towards adding some features, i haven't encountered any issues with the functionality added with this PR.
I'm perfectly fine with working upstream on the changes made.

Unless anyone requests any code / style changes, feel free to merge.

bulzipke commented:
@twinaphex @MrHuu 
Thank you for your consideration.

I have no plans for improvement in this PR right now.
I think it will be good to merge with Mr Huu's wonderful function in other PR.

And the png resources used by this PR are listed below.
It has to be done together.

https://github.com/libretro/retroarch-assets/pull/408

Thank you :)

MarioKartFan commented:
@twinaphex Can you let us know what is keeping this PR from being merged? Thanks!

MrHuu commented:
Alright, i'm able to start working on this again.
I do have a few questions i'd like to discuss about the actual functionality regarding this PR and the functionality i'm working towards.

Right now this PR works slightly different from what i have available.

As it stand now ( this PR ), the bottom screen is idle while running a game. When input is detected, the core is paused and the menu is toggled. The entire screen is used to have 3 options: resume game, save state and load state. When a button is pressed, the menu is toggled again and the core is resumed.

This however conflicts with the functionality i'm working on. As indicated by @bulzipke plan A.
To make this a reality, new assets need to be provided to allow the bottom selection buttons to co-exist.

Also the save-sate functionality i had, did not toggle the menu nor paused the core. The bottom middle 'Retroarch menu' button is always available to toggle the menu. But not necessary for the save-sate functionality.
Right now, the 'Resume game' button and the 'Retroarch menu' button conflict with each other.

I don't want to change @bulzipke 's work without any discussion. I don't feel i'm the one to decide on my own.

Does anyone want to make a suggestion about the layout or functionality?
I was personally thinking of maybe replacing the 'Resume game' button with an option to increase or decrease the selected save slot?

MarioKartFan commented:
@MrHuu Reset Game would be useful and aesthetically consistent with the Virtual Console menu.

bulzipke commented:
I think don't need a resume game button for menu toggle too.
Because users have to touch twice every time for save state or load state, it seems that can give the user an uncomfortable experience.

Replacing the 'Resume game' button with an option to increase or decrease the selected save slot seems great idea!

MrHuu commented:
![savestate](https://user-images.githubusercontent.com/561623/130660352-e9bed111-16d1-4398-ae76-c5b23862830e.png)

This is where i'm currently at. Sorry @bulzipke 
Switching save-slots has to wait for now, i really want to be able put out a proper pr first.

MarioKartFan commented:
This looks absolutely beautiful.

MrHuu commented:
I'm still trying to work out a few issues. Which may cause some delay.
While everything seems to be running great with most cores, the more demanding cores like PCSX-ReARMed and Dosbox-svn either don't run at all or crash on some occasions.
But that's not related to this PR.

@twinaphex
Could this PR be merged?
I've been able to test out a lot of cores without any issues.

The added ability to save and load states to and from RAM during runtime is a great addition.
Especially with savestates larger in size, saving and loading is instant.

A notification when both saving and loading states to RAM does need be be added.
And probably also when saving the RAM state to a file. Although this may conflict with the command also being executed when the video driver is de-initialized.
```
CMD_EVENT_SAVE_STATE_TO_RAM
CMD_EVENT_LOAD_STATE_FROM_RAM
CMD_EVENT_RAM_STATE_TO_FILE
```
The bottom screen menu is fully functional, and does a great job on it's initial design.

I would really love to see this merged, to be tested out and enjoyed by a greater public.
And i'm starting to feel bad for @bulzipke, his work needs to be appreciated.

Great PR with much appreciated functionality !

MarioKartFan commented:
Would be nice to see this merged or else understand whether there are technical concerns.

twinaphex commented:
This looks great indeed. I will check back again with @jdgleaver on this one and we'll see how to proceed with this one.

jdgleaver commented:
@bulzipke Please accept my apologies. This PR fell entirely off my radar, and I missed all the discussions in this thread....

There is only one blocking issue in your code, and it's this function here:

https://github.com/libretro/RetroArch/pull/12470/files#diff-4dd6e477da06a009ecb1c94f3eb145e8655b78879bc1a47d349183ca02092a2cR9286-R9305

In `path_get_state(void)` you use and return a static variable: `state_path`. We cannot do it like this. Global state can only exist in specific structs, and the dynamic nature of the save state path (i.e. dependence on index) makes the use of a static variable here very thread unsafe.

What I would suggest is that you refactor the code as follows:

- Remove the `RARCH_PATH_STATE` enum
- Rewrite `path_get_state(void)` as something like:

```c
bool retroarch_get_current_savestate_path(char *path, size_t len)
{
   struct rarch_state *p_rarch = &rarch_st;
   const global_t *global      = &p_rarch->g_extern;
   settings_t *settings        = p_rarch->configuration_settings;
   int state_slot              = settings ? settings->ints.state_slot : 0;
   const char *name_savestate  = NULL;

   if (!path || !global)
      return false;

   name_savestate = global->name.savestate;
   if (string_is_empty(name_savestate))
      return false;

   if (state_slot > 0)
      snprintf(path, len, "%s%d",  name_savestate, state_slot);
   else if (state_slot < 0)
      fill_pathname_join_delim(path, name_savestate, "auto", '.', len);
   else
      strlcpy(path, name_savestate, len);

   return true;
}
```

- Add `retroarch_get_current_savestate_path()` to `retroarch.h`

- Call `retroarch_get_current_savestate_path()` directly in `command_event_main_state()` and `ctr_texture_path()`

If you can make these small changes, I think we can go ahead and merge this directly (after a squash+rebase)

(One other comment: It would be nice if the `Tap the Touch Screen to go to the Retroarch menu` text were to dissapear after a period of inactivity following load content - but this could be done in a separate PR)

bulzipke commented:
@jdgleaver Thank you for your kind guidance.
I modified it as you advised.

Please let me know if you need anything else :)

jdgleaver commented on retroarch.c:
```diff
@@ -11449,22 +11474,9 @@ static bool command_event_main_state(
-   {
-      int state_slot             = settings->ints.state_slot;
-      const char *name_savestate = global->name.savestate;
+   msg[0]                      = '\0';
```
@bulzipke Many thanks for updating the PR! There's just one tiny issue remaining. Because you changed:

```
state_path[0] = msg[0]      = '\0';
```
to
```
msg[0]                      = '\0';
```
`state_path` will end up unterminated in the event that `retroarch_get_current_savestate_path()` fails. In this case, anything that uses `state_path` will subsequently overflow the buffer.

So basically, in both the places that you call `retroarch_get_current_savestate_path()`, you just need to make sure that `state_path[0] = '\0'` is set first.

With this last fix, we can go ahead and merge :)

bulzipke replied:
Oh, thank you for pointing out my mistake.
I'm still unfamiliar with github, so please let me know if there's any additional work to do (e.g., rebase, etc.)

twinaphex commented:
Squash and merged it.

bulzipke commented:
@jdgleaver @twinaphex 
I always pay tribute to your efforts.
Thank you, RetroArch!

twinaphex commented:
@bulzipke Very glad to finally have this merged and for it to be ready for 1.9.9 this weekend.

I'd love to see you and @MrHuu further hone and improve this from this point on. Definitely also keep performance in mind in future PRs to see if there is stuff we can do there. Is it also possible to turn this off with a setting BTW, or is it always on?

bulzipke commented:
@twinaphex 
User can turn the screen on and off with the following settings:
Settings->User Interface->3DS Bottom Screen

And although this option will not affect performance, it is estimated that "Tap the Touch Screen to go to the Retroarch menu" text may be burn-in (not sure if they occurs to 3DS) or additional battery consumption to draw the screen.

twinaphex commented:
Thanks for that clarification. I  will add that to the blog post

MarioKartFan commented:
Just a thought: I would be hesitant about flagging risk of burn-in. Having been around the scene for years and knowing that some people log hundreds of hours with games that have static bottom screens, I’m not aware of burn-in being an issue for the 3DS.
